scenarios/pfcp-app-filtering.xml

```diff
@@ -3,56 +3,98 @@
-                <step name="Uplink-Enb-Send-Gtp-App-Filtering-${#}" requires="~Check-Traffic-App-Filtering-Negative" delay="5"
+                <step name="Uplink-Enb-Send-Gtp-App-Filtering-${#}-1" requires="~Check-Traffic-App-Filtering-Negative-1" delay="5"
                       exec="mn-cmd ${ENODEB#} traffic.py send-gtp -c 10 --flow-count 5 --teid-base ${#}0 --ue-pool 17.0.${#-1}.0/24 --enb-addr 140.0.10${#-1}.1 --pdn-port 81"/>
             </parallel>
         </group>
-        <group name="Pfcp-Clear-App-Filtering">
+        <group name="Pfcp-Clear-App-Filtering-1">
             <sequential var="${ENODEB#}"
-                starts="Clear-App-Filtering-${#}"
-                ends="Clear-App-Filtering-${#-1}">
-                <step name="Clear-App-Filtering-${#}" requires="~Check-Traffic-App-Filtering-Negative,~Check-Traffic-App-Filtering-Positive"
+                        starts="Clear-App-Filtering-${#}-1"
+                        ends="Clear-App-Filtering-${#-1}-1">
+                <step name="Clear-App-Filtering-${#}-1" requires="~Check-Traffic-App-Filtering-Negative-1,~Check-Traffic-App-Filtering-Positive-1"
+                      exec="${DOCKER_COMPOSE_CMD} exec -T mock-smf pfcpctl session delete --count 5 --baseID ${#}0"/>
+            </sequential>
+        </group>
+        <group name="Pfcp-Setup-App-Filtering-2" requires="Pfcp-Clear-App-Filtering-1">
+            <group name="Pfcp-Push-App-Filtering-2" delay="5">
+                <sequential var="${ENODEB#}"
+                            starts="Pfcp-Session-Create-App-Filtering-${#}-2"
+                            ends="Pfcp-Session-App-Filtering-${#-1}-2">
+                    <step name="Pfcp-Session-Create-App-Filtering-${#}-2"
+                          exec="${DOCKER_COMPOSE_CMD} exec -T mock-smf pfcpctl session create --count 5 --baseID ${#}0 --ue-pool 17.0.${#-1}.0/24 --gnb-addr 140.0.10${#-1}.1 --sdf-filter 'permit out ip from any to assigned' --gate-status"/>
+                    <step name="Pfcp-Session-App-Filtering-${#}-2" requires="Pfcp-Session-Create-App-Filtering-${#}-2"
+                          exec="${DOCKER_COMPOSE_CMD} exec -T mock-smf pfcpctl session modify --count 5 --baseID ${#}0 --ue-pool 17.0.${#-1}.0/24 --gnb-addr 140.0.10${#-1}.1"/>
+                </sequential>
+            </group>
+            <step name="Check-Up4-Flows-App-Filtering-2" requires="Pfcp-Push-App-Filtering-2" delay="5"
+                  exec="onos-cli-grep ${OCI} up4:read-flows 'Apps=1, UL sess=${PFCP_UP4_FLOWS}, UL flows=${PFCP_UP4_FLOWS}, DL flows=${PFCP_UP4_FLOWS}'"/>
+            <step name="Check-Flow-Rules-App-Filtering-2" requires="Pfcp-Push-App-Filtering-2" delay="5"
+                  exec="onos-check-flows ${OCI}"/>
+        </group>
+        <group name="Check-Traffic-App-Filtering-Negative-2" requires="Pfcp-Setup-App-Filtering-2" delay="5">
+            <parallel var="${ENODEB#}">
```

Comment:
is it not dangerous doing these steps in parallel ?

Reply:
Why do you think is dangerous? We send traffic and check if we didn't receive any packets.

Reply:
I don't think you have any guarantee about the ordering - what about if you send first the traffic and then check the rcv ? You are not actually testing the behavior you expect - no ?

Reply:
The steps sending traffic have a delay of 5 sec, this should save us from missing packets.